// @dbos-inc/dbos-sdk/dist/src/telemetry/traces.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tracer = void 0;
const sdk_trace_base_1 = require("@opentelemetry/sdk-trace-base");
const resources_1 = require("@opentelemetry/resources");
const api_1 = __importDefault(require("@opentelemetry/api"));
const semantic_conventions_1 = require("@opentelemetry/semantic-conventions");
const core_1 = require("@opentelemetry/core");
class Tracer {
    telemetryCollector;
    tracer;
    applicationID;
    applicationVersion;
    executorID;
    constructor(telemetryCollector) {
        this.telemetryCollector = telemetryCollector;
        this.tracer = new sdk_trace_base_1.BasicTracerProvider({
            resource: new resources_1.Resource({
                [semantic_conventions_1.SemanticResourceAttributes.SERVICE_NAME]: "dbos",
            }),
        });
        this.tracer.register();
        this.applicationID = process.env.DBOS__APPID || "";
        this.applicationVersion = process.env.DBOS__APPVERSION || "";
        this.executorID = process.env.DBOS__VMID || "local"; // for consistency with src/context.ts
    }
    startSpanWithContext(spanContext, name, attributes) {
        const tracer = api_1.default.trace.getTracer("dbos-tracer");
        const ctx = api_1.default.trace.setSpanContext(api_1.default.context.active(), spanContext);
        return tracer.startSpan(name, { startTime: performance.now(), attributes: attributes }, ctx);
    }
    startSpan(name, attributes, parentSpan) {
        const tracer = api_1.default.trace.getTracer("dbos-tracer");
        const startTime = (0, core_1.hrTime)(performance.now());
        if (parentSpan) {
            const ctx = api_1.default.trace.setSpan(api_1.default.context.active(), parentSpan);
            return tracer.startSpan(name, { startTime: startTime, attributes: attributes }, ctx);
        }
        else {
            return tracer.startSpan(name, { startTime: startTime, attributes: attributes });
        }
    }
    endSpan(span) {
        span.end((0, core_1.hrTime)(performance.now()));
        span.attributes.applicationID = this.applicationID;
        span.attributes.applicationVersion = this.applicationVersion;
        if (!("executorID" in span.attributes)) {
            span.attributes.executorID = this.executorID;
        }
        this.telemetryCollector.push(span);
    }
}
exports.Tracer = Tracer;
