// @dbos-inc/dbos-sdk/dist/src/dbos-runtime/runtime.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DBOSRuntime = exports.defaultEntryPoint = void 0;
const dbos_executor_1 = require("../dbos-executor");
const server_1 = require("../httpServer/server");
const fs = __importStar(require("fs"));
const lodash_1 = require("lodash");
const error_1 = require("../error");
const node_path_1 = __importDefault(require("node:path"));
const url_1 = require("url");
const scheduler_1 = require("../scheduler/scheduler");
const wfqueue_1 = require("../wfqueue");
const dbos_1 = require("../dbos");
exports.defaultEntryPoint = "dist/operations.js";
class DBOSRuntime {
    runtimeConfig;
    dbosConfig;
    dbosExec = undefined;
    servers;
    scheduler = undefined;
    wfQueueRunner = undefined;
    constructor(dbosConfig, runtimeConfig) {
        this.runtimeConfig = runtimeConfig;
        // Initialize workflow executor.
        this.dbosConfig = dbosConfig;
        dbos_1.DBOS.dbosConfig = dbosConfig;
    }
    /**
     * Initialize the runtime and start the server
     */
    async initAndStart() {
        try {
            this.dbosExec = new dbos_executor_1.DBOSExecutor(this.dbosConfig);
            dbos_1.DBOS.globalLogger = this.dbosExec.logger;
            this.dbosExec.logger.debug(`Loading classes from entrypoints ${JSON.stringify(this.runtimeConfig.entrypoints)}`);
            await DBOSRuntime.loadClasses(this.runtimeConfig.entrypoints);
            await this.dbosExec.init();
            const server = new server_1.DBOSHttpServer(this.dbosExec);
            this.servers = await server.listen(this.runtimeConfig.port, this.runtimeConfig.admin_port);
            this.dbosExec.logRegisteredHTTPUrls();
            this.scheduler = new scheduler_1.DBOSScheduler(this.dbosExec);
            this.scheduler.initScheduler();
            this.scheduler.logRegisteredSchedulerEndpoints();
            wfqueue_1.wfQueueRunner.logRegisteredEndpoints(this.dbosExec);
            this.wfQueueRunner = wfqueue_1.wfQueueRunner.dispatchLoop(this.dbosExec);
            for (const evtRcvr of this.dbosExec.eventReceivers) {
                await evtRcvr.initialize(this.dbosExec);
            }
            for (const evtRcvr of this.dbosExec.eventReceivers) {
                evtRcvr.logRegisteredEndpoints();
            }
        }
        catch (error) {
            this.dbosExec?.logger.error(error);
            if (error instanceof error_1.DBOSFailLoadOperationsError) {
                console.error('\x1b[31m%s\x1b[0m', "Did you compile this application? Hint: run `npm run build` and try again");
                process.exit(1);
            }
            await this.destroy(); //wrap up, i.e. flush log contents to OpenTelemetry exporters
            process.exit(1);
        }
        const onSigterm = this.onSigterm.bind(this);
        process.on('SIGTERM', onSigterm);
        process.on('SIGQUIT', onSigterm);
    }
    /**
     * Load an application's workflow functions, assumed to be in src/operations.ts (which is compiled to dist/operations.js).
     */
    static async loadClasses(entrypoints) {
        const allClasses = [];
        for (const entrypoint of entrypoints) {
            const operations = node_path_1.default.isAbsolute(entrypoint) ? entrypoint : node_path_1.default.join(process.cwd(), entrypoint);
            let exports;
            if (fs.existsSync(operations)) {
                const operationsURL = (0, url_1.pathToFileURL)(operations).href;
                exports = (await import(operationsURL));
            }
            else {
                throw new error_1.DBOSFailLoadOperationsError(`Failed to load operations from the entrypoint ${entrypoint}`);
            }
            const classes = [];
            for (const key in exports) {
                if ((0, lodash_1.isObject)(exports[key])) {
                    classes.push(exports[key]);
                }
            }
            allClasses.push(...classes);
        }
        if (allClasses.length === 0) {
            throw new error_1.DBOSFailLoadOperationsError("operations not found");
        }
        return allClasses;
    }
    onSigterm() {
        this.dbosExec?.logger.info("Stopping application: received a termination signal");
        void this.destroy().finally(() => {
            process.exit(1);
        });
    }
    /**
     * Shut down the HTTP and other services and destroy workflow executor.
     */
    async destroy() {
        await this.scheduler?.destroyScheduler();
        try {
            wfqueue_1.wfQueueRunner.stop();
            await this.wfQueueRunner;
        }
        catch (err) {
            const e = err;
            this.dbosExec?.logger.warn(`Error destroying workflow queue runner: ${e.message}`);
        }
        for (const evtRcvr of this.dbosExec?.eventReceivers || []) {
            await evtRcvr.destroy();
        }
        if (this.servers) {
            this.servers.appServer?.close();
            this.servers.adminServer.close();
        }
        await this.dbosExec?.destroy();
    }
}
exports.DBOSRuntime = DBOSRuntime;
