// @dbos-inc/dbos-sdk/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DBOS = exports.WorkflowQueue = exports.parseConfigFile = exports.Scheduled = exports.SchedulerConfig = exports.SchedulerMode = exports.KafkaConsume = exports.Kafka = exports.OpenApiSecurityScheme = exports.KoaMiddleware = exports.KoaGlobalMiddleware = exports.KoaCors = exports.KoaBodyParser = exports.Authentication = exports.DeleteApi = exports.PutApi = exports.PatchApi = exports.PostApi = exports.GetApi = exports.ArgSource = exports.ArgSources = exports.associateClassWithEventReceiver = exports.associateMethodWithEventReceiver = exports.DBOSDeploy = exports.DBOSInitializer = exports.RequiredRole = exports.StoredProcedure = exports.Communicator = exports.Step = exports.Workflow = exports.Transaction = exports.configureInstance = exports.ConfiguredInstance = exports.OrmEntities = exports.DefaultArgOptional = exports.DefaultArgRequired = exports.DefaultRequiredRole = exports.ArgVarchar = exports.ArgDate = exports.ArgName = exports.LogMask = exports.SkipLogging = exports.ArgOptional = exports.ArgRequired = exports.LogMasks = exports.DBOSResponseError = exports.Error = exports.StatusString = exports.InitContext = exports.createTestingRuntime = void 0;
var testing_runtime_1 = require("./testing/testing_runtime");
Object.defineProperty(exports, "createTestingRuntime", { enumerable: true, get: function () { return testing_runtime_1.createTestingRuntime; } });
var context_1 = require("./context");
Object.defineProperty(exports, "InitContext", { enumerable: true, get: function () { return context_1.InitContext; } });
var workflow_1 = require("./workflow");
Object.defineProperty(exports, "StatusString", { enumerable: true, get: function () { return workflow_1.StatusString; } });
exports.Error = __importStar(require("./error"));
var error_1 = require("./error");
Object.defineProperty(exports, "DBOSResponseError", { enumerable: true, get: function () { return error_1.DBOSResponseError; } });
var decorators_1 = require("./decorators");
Object.defineProperty(exports, "LogMasks", { enumerable: true, get: function () { return decorators_1.LogMasks; } });
// Parameter Decorators
Object.defineProperty(exports, "ArgRequired", { enumerable: true, get: function () { return decorators_1.ArgRequired; } });
Object.defineProperty(exports, "ArgOptional", { enumerable: true, get: function () { return decorators_1.ArgOptional; } });
Object.defineProperty(exports, "SkipLogging", { enumerable: true, get: function () { return decorators_1.SkipLogging; } });
Object.defineProperty(exports, "LogMask", { enumerable: true, get: function () { return decorators_1.LogMask; } });
Object.defineProperty(exports, "ArgName", { enumerable: true, get: function () { return decorators_1.ArgName; } });
Object.defineProperty(exports, "ArgDate", { enumerable: true, get: function () { return decorators_1.ArgDate; } });
Object.defineProperty(exports, "ArgVarchar", { enumerable: true, get: function () { return decorators_1.ArgVarchar; } });
// Class Decorators
Object.defineProperty(exports, "DefaultRequiredRole", { enumerable: true, get: function () { return decorators_1.DefaultRequiredRole; } });
Object.defineProperty(exports, "DefaultArgRequired", { enumerable: true, get: function () { return decorators_1.DefaultArgRequired; } });
Object.defineProperty(exports, "DefaultArgOptional", { enumerable: true, get: function () { return decorators_1.DefaultArgOptional; } });
// Typeorm Class Decorators
Object.defineProperty(exports, "OrmEntities", { enumerable: true, get: function () { return decorators_1.OrmEntities; } });
// Class Instances
Object.defineProperty(exports, "ConfiguredInstance", { enumerable: true, get: function () { return decorators_1.ConfiguredInstance; } });
Object.defineProperty(exports, "configureInstance", { enumerable: true, get: function () { return decorators_1.configureInstance; } });
// Method Decorators
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return decorators_1.Transaction; } });
Object.defineProperty(exports, "Workflow", { enumerable: true, get: function () { return decorators_1.Workflow; } });
Object.defineProperty(exports, "Step", { enumerable: true, get: function () { return decorators_1.Step; } });
Object.defineProperty(exports, "Communicator", { enumerable: true, get: function () { return decorators_1.Step; } });
Object.defineProperty(exports, "StoredProcedure", { enumerable: true, get: function () { return decorators_1.StoredProcedure; } });
Object.defineProperty(exports, "RequiredRole", { enumerable: true, get: function () { return decorators_1.RequiredRole; } });
Object.defineProperty(exports, "DBOSInitializer", { enumerable: true, get: function () { return decorators_1.DBOSInitializer; } });
Object.defineProperty(exports, "DBOSDeploy", { enumerable: true, get: function () { return decorators_1.DBOSDeploy; } });
// Extensions for others to register event receivers/pollers
Object.defineProperty(exports, "associateMethodWithEventReceiver", { enumerable: true, get: function () { return decorators_1.associateMethodWithEventReceiver; } });
Object.defineProperty(exports, "associateClassWithEventReceiver", { enumerable: true, get: function () { return decorators_1.associateClassWithEventReceiver; } });
var handlerTypes_1 = require("./httpServer/handlerTypes");
Object.defineProperty(exports, "ArgSources", { enumerable: true, get: function () { return handlerTypes_1.ArgSources; } });
var handler_1 = require("./httpServer/handler");
// Endpoint Parameter Decorators
Object.defineProperty(exports, "ArgSource", { enumerable: true, get: function () { return handler_1.ArgSource; } });
// Endpoint Decorators
Object.defineProperty(exports, "GetApi", { enumerable: true, get: function () { return handler_1.GetApi; } });
Object.defineProperty(exports, "PostApi", { enumerable: true, get: function () { return handler_1.PostApi; } });
Object.defineProperty(exports, "PatchApi", { enumerable: true, get: function () { return handler_1.PatchApi; } });
Object.defineProperty(exports, "PutApi", { enumerable: true, get: function () { return handler_1.PutApi; } });
Object.defineProperty(exports, "DeleteApi", { enumerable: true, get: function () { return handler_1.DeleteApi; } });
var middleware_1 = require("./httpServer/middleware");
// Middleware Decorators
Object.defineProperty(exports, "Authentication", { enumerable: true, get: function () { return middleware_1.Authentication; } });
Object.defineProperty(exports, "KoaBodyParser", { enumerable: true, get: function () { return middleware_1.KoaBodyParser; } });
Object.defineProperty(exports, "KoaCors", { enumerable: true, get: function () { return middleware_1.KoaCors; } });
Object.defineProperty(exports, "KoaGlobalMiddleware", { enumerable: true, get: function () { return middleware_1.KoaGlobalMiddleware; } });
Object.defineProperty(exports, "KoaMiddleware", { enumerable: true, get: function () { return middleware_1.KoaMiddleware; } });
// OpenApi Decorators
Object.defineProperty(exports, "OpenApiSecurityScheme", { enumerable: true, get: function () { return middleware_1.OpenApiSecurityScheme; } });
var kafka_1 = require("./kafka/kafka");
Object.defineProperty(exports, "Kafka", { enumerable: true, get: function () { return kafka_1.Kafka; } });
Object.defineProperty(exports, "KafkaConsume", { enumerable: true, get: function () { return kafka_1.KafkaConsume; } });
var scheduler_1 = require("./scheduler/scheduler");
Object.defineProperty(exports, "SchedulerMode", { enumerable: true, get: function () { return scheduler_1.SchedulerMode; } });
Object.defineProperty(exports, "SchedulerConfig", { enumerable: true, get: function () { return scheduler_1.SchedulerConfig; } });
Object.defineProperty(exports, "Scheduled", { enumerable: true, get: function () { return scheduler_1.Scheduled; } });
var config_1 = require("./dbos-runtime/config");
Object.defineProperty(exports, "parseConfigFile", { enumerable: true, get: function () { return config_1.parseConfigFile; } });
var wfqueue_1 = require("./wfqueue");
Object.defineProperty(exports, "WorkflowQueue", { enumerable: true, get: function () { return wfqueue_1.WorkflowQueue; } });
var dbos_1 = require("./dbos");
Object.defineProperty(exports, "DBOS", { enumerable: true, get: function () { return dbos_1.DBOS; } });
